### PR TITLE
 plugin: support dynamic enable/disable plugins

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

- support dynamic disable/enable a loaded plugins.
- have ablity to  observe plugin's enable/disable state
- enable/disable operation will affect all nodes in the cluster

### What is changed and how it works?

need use parser's new syntax https://github.com/pingcap/parser/pull/374

then just change current disable flag and broadcast change via etcd

attention: we also need update plugins to let plugin's go.mod dep on new parser...

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

```
./bin/tidb-server -store tikv -path 127.0.0.1:2379 -plugin-dir=$GOPATH/src/github.com/pingcap/enterprise-plugin/audit -plugin-load=audit-1

mysql> show plugins;
+-------+--------------+-------+------------------------------------------------------------------------------+---------+---------+
| Name  | Status       | Type  | Library                                                                      | License | Version |
+-------+--------------+-------+------------------------------------------------------------------------------+---------+---------+
| audit | Ready-enable | Audit | /home/robi/code/go/src/github.com/pingcap/enterprise-plugin/audit/audit-1.so |         | 1       |
+-------+--------------+-------+------------------------------------------------------------------------------+---------+---------+
1 row in set (0.01 sec)

mysql> admin plugins disable audit;
Query OK, 0 rows affected (0.01 sec)

mysql> show plugins;
+-------+---------------+-------+------------------------------------------------------------------------------+---------+---------+
| Name  | Status        | Type  | Library                                                                      | License | Version |
+-------+---------------+-------+------------------------------------------------------------------------------+---------+---------+
| audit | Ready-disable | Audit | /home/robi/code/go/src/github.com/pingcap/enterprise-plugin/audit/audit-1.so |         | 1       |
+-------+---------------+-------+------------------------------------------------------------------------------+---------+---------+
1 row in set (0.00 sec)

mysql> admin plugins enable audit;
Query OK, 0 rows affected (0.00 sec)

mysql> show plugins;
+-------+--------------+-------+------------------------------------------------------------------------------+---------+---------+
| Name  | Status       | Type  | Library                                                                      | License | Version |
+-------+--------------+-------+------------------------------------------------------------------------------+---------+---------+
| audit | Ready-enable | Audit | /home/robi/code/go/src/github.com/pingcap/enterprise-plugin/audit/audit-1.so |         | 1       |
+-------+--------------+-------+------------------------------------------------------------------------------+---------+---------+
1 row in set (0.00 sec)
```

Code changes

 - impl change

Side effects

 - N/A

Related changes

 - Need to cherry-pick to the release 3.0/2.1

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/11122)

### CODE REVIEW

**Comment by lysu** [comment]
plugin build CI will be fixed after https://github.com/pingcap/enterprise-plugin/pull/8 be merged

File: plugin/plugin.go [hunk 1]
```diff
@@ -404,7 +451,25 @@ func NotifyFlush(dom *domain.Domain, pluginName string) error {
...
 	}
-	_, err := dom.GetEtcdClient().KV.Put(context.Background(), p.Manifest.flushWatcher.path, "")
+	_, err := dom.GetEtcdClient().KV.Put(context.Background(), p.Manifest.flushWatcher.path, strconv.Itoa(int(p.Disabled)))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ChangeDisableFlagAndFlush changes plugin disable flag and notify other nodes to do same change..
```
Comment (tiancaiamao) [inline plugin/plugin.go]: s/../.

File: plugin/plugin.go [hunk 2]
```diff
@@ -404,7 +451,25 @@ func NotifyFlush(dom *domain.Domain, pluginName string) error {
...
+func ChangeDisableFlagAndFlush(dom *domain.Domain, pluginName string, disable bool) error {
+	p := getByName(pluginName)
+	disableInt := uint32(0)
+	if disable {
+		disableInt = 1
+	}
+	atomic.StoreUint32(&p.Disabled, disableInt)
+	if p == nil || p.Manifest.flushWatcher == nil || p.State != Ready {
```
Comment (crazycs520) [inline plugin/plugin.go]: Should move `if p == nil` after line#463 to avoid nil point panic in line#468 ?

Comment (lysu) [reply plugin/plugin.go]: done~~

**Pull request merged**